extensions/amp-story-360/0.1/amp-story-360.js

```diff
@@ -0,0 +1,291 @@
+        return null;
+      } else {
+        return this.orientations[this.currentHeadingIndex];
+      }
+    }
+    const easing = this.easeInOutQuad_(
+        this.currentFrame % framesPerSection / framesPerSection);
+    const from = this.orientations[this.currentHeadingIndex];
+    const to = this.orientations[this.currentHeadingIndex + 1];
+    return new CameraOrientation(
+        from.theta + (to.theta - from.theta) * easing,
+        from.phi + (to.phi - from.phi) * easing,
+        from.scale + (to.scale - from.scale) * easing);
+  }
+}
+
+export class AmpStory360 extends AMP.BaseElement {
+  /** @param {!AmpElement} element */
+  constructor(element) {
+    super(element);
+
+    /** @private {!Array<!CameraOrientation>} */
+    this.orientations_ = [];
+
+    /** @private {?number} */
+    this.duration_ = null;
+
+    /** @private {?Element} */
+    this.container_ = null;
+
+    /** @private {?Renderer} */
+    this.renderer_ = null;
+
+    /** @private {?CameraAnimation} */
+    this.animation_ = null;
+
+    /** @private {boolean} */
+    this.isPlaying_ = false;
+  }
+
+  /** @override */
+  buildCallback() {
+    const attr = (name) => this.element.getAttribute(name);
+
+    this.duration_ =
+        attr('duration') ? timeStrToMillis(attr('duration')) : null;
+
+    const startHeading = parseFloat(attr('heading-start') || 0);
+    const startPitch = parseFloat(attr('pitch-start') || 0);
+    const startZoom = parseFloat(attr('zoom-start') || 1);
+    this.orientations_.push(
+        CameraOrientation.fromDegrees(startHeading, startPitch, startZoom));
+
+    if (attr('heading-end') !== undefined || attr('pitch-end') !== undefined ||
+        attr('zoom-end') !== undefined) {
+      const endHeading = parseFloat(attr('heading-end') || 0);
+      const endPitch = parseFloat(attr('pitch-end') || 0);
+      const endZoom = parseFloat(attr('zoom-end') || 1);
+      this.orientations_.push(
+          CameraOrientation.fromDegrees(endHeading, endPitch, endZoom));
+    }
+
+    this.container_ = this.element.ownerDocument.createElement('div');
+    const canvas = this.element.ownerDocument.createElement('canvas');
+    this.element.appendChild(this.container_);
+    this.container_.appendChild(canvas);
+    this.renderer_ = new Renderer(canvas);
+    this.applyFillContent(this.container_, /* replacedContent */ true);
+  }
+
+  /** @override */
+  isLayoutSupported(layout) {
+    return isLayoutSizeDefined(layout);
+  }
+
+  /** @override */
+  layoutCallback() {
+    const ampImgEl = this.element.querySelector('amp-img');
+    userAssert(ampImgEl, 'amp-story-360 must contain an amp-img element.');
+    const owners = Services.ownersForDoc(this.element);
+    owners.setOwner(ampImgEl, this.element);
+    owners.scheduleLayout(this.element, ampImgEl);
+    return whenUpgradedToCustomElement(ampImgEl)
+        .then(() => {
+          return ampImgEl.signals().whenSignal(CommonSignals.LOAD_END);
```

Comment:
This one can fail if the image doesn't load. Can you catch the rejected promise and log an error? eg: https://github.com/ampproject/amphtml/blob/master/src/experiments.js#L240